Create experimental SelectTree component

nathanss opened the pull request:
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This creates the experimental TreeSelect control along with a storybook example, which will be used through the new product editing experience, for example at the Category field.


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Please include detailed instructions on how these changes can be tested, make sure to review and follow the guide for writing high-quality testing instructions below. -->

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

1. Run storybook: `pnpm --filter="storybook" storybook`
2. Look for the newly added story ([link](http://localhost:6007/?path=/story/woocommerce-admin-experimental-selecttreecontrol--multiple-select-tree))
3. Test it out

<!-- End testing instructions -->

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [ ] Have you included testing instructions?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

nathanss committed:
Create a new experimental component TreeSelect
Add functionality to TreeControl to support additional scenarios

nathanss committed:
Fix a bug where the shouldNotRecursivelySelect was not passed to Trees inside Tree Items

nathanss committed:
Do not collapse nodes inside useEffect if they are already expanded

nathanss committed:
Add storybook example for SelectTreeControl

nathanss committed:
Move component to its own directory

nathanss committed:
Remove wrong comment

nathanss committed:
Fix CSS import path

nathanss commented:
The E2E tests are also breaking in trunk (I tested). It has nothing to do with the PR.

mdperez86 commented on packages/js/components/src/experimental-select-tree-control/stories/index.tsx:
```diff
@@ -0,0 +1,92 @@
+			multiple
+			items={ listItems }
+			selected={ selected }
+			getFilteredItems={ ( items, searchValue ) => {
```
We should not have this filter here. We should use [this](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/components/src/experimental-select-control/hooks/use-async-filter.tsx) instead for filtering items.

nathanss replied:
I'm wondering how it will look like on the CategoryField implementation. Can we leave that to the follow-up PR that implements it on the CategoryField?

mdperez86 commented on packages/js/components/src/experimental-select-tree-control/select-tree.tsx:
```diff
@@ -0,0 +1,221 @@
+	getItemLabel = ( item ) => item?.name || '',
+	getItemValue = ( item ) => item?.id || '',
+	getSelectedItemProps,
+	getFilteredItems,
```
This function introduces another layer of complexity inside the component at render time. What `items` prop is for then? Filtering items should be a responsibility managed outside of the component like in the server.

nathanss replied:
I changed the implementation. One upside of always having all the items inside SelectTree and Tree controls would be to implement reusable logic, but I think it's fine to switch it for now.

mdperez86 commented on packages/js/components/src/experimental-select-tree-control/select-tree.tsx:
```diff
@@ -0,0 +1,221 @@
+								}
+								setIsFocused( true );
+							},
+							onBlur: ( event ) => {
```
I would move all these functionalities to a custom hook like `useCombobox` or something similar.

nathanss replied:
Since this is inside a .map, it gets trickier to create a custom hook. Also, I prefer leaving it like this, since the component is not so big and in case we need the same props in two different places, then we can create the hook.

mdperez86 commented on packages/js/components/src/experimental-tree-control/types.ts:
```diff
@@ -1,7 +1,7 @@
 	parent?: string;
-	value: string;
-	label: string;
+	id: string;
```
Instead of changing these props why you didn't do generic here. Like in the `select-control`, `select-tree-control` and `selected-items` components, at the end you make use of the following function anyway because of that right. 
```typescript
getItemLabel?: getItemLabelType< Item >;
getItemValue?: getItemValueType< Item >;
```
The problem with generic types is that you need to let the component knows how to get access to the `label` and `value` props within the passed object right. But you already have that! So we can apply the same to the Tree so we can pass a generic type instead of fixed type (`Item`).
The advantage of this is that we can pass any object list down to the tree.

nathanss replied:
I decided to rename it back and not use generics for now. You're correct that it didn't make sense having getItemLabel and getItemValue as props on the component the way it previously was, so I removed it. It would be a considerate effort to implement generics in the Tree control because the `Item` type is heavily used throughout the component, for example in `useLinkedTree` and `useSelection`

mdperez86 reviewed (changes requested):
In general the behavior looks correct to me. Giant job here @nathanss !!!
I left some comments

nathanss committed:
Remove getFilteredItems from select-tree

nathanss committed:
Stop using getFilteredItems inside select tree
Create new shouldShowCreateButton prop to allow receiving only filtered items inside component
Remove support to show custom getItemLabel and getItemValue

nathanss committed:
Change interface

nathanss committed:
Rename back Item type names

nathanss committed:
Rename back value

nathanss committed:
Fix lint error and remove passing ref to treeProps

mdperez86 commented on packages/js/components/src/experimental-select-tree-control/stories/index.tsx:
```diff
@@ -28,35 +28,37 @@ const listItems: Item[] = [
 	{ id: '3', name: 'Other' },
 ];
 
+const getFilteredItems = ( items, searchValue ) => {
```
This is in theory the same thing as having it inside the SelectTree. So if you are not going to remove this function in this PR then we should mark the function as something to be removed in the future.

mdperez86 commented on packages/js/components/src/experimental-tree-control/tree.tsx:
```diff
@@ -50,13 +46,11 @@ export const Tree = forwardRef( function ForwardedTree(
-								 )?.focus();
-							}
+							(
+								document.querySelector(
```
What happen if we have multiple Tree components rendered in the same page? Consider that the Tree component can be used not only inside a Popover.

nathanss replied:
I'm having a hard time improving this, because the tree is recursive (TreeItems can have Trees), so I can't know the exact DOM structure. It would help a lot if I could get the ref to the button, but that's not working. Any ideas?

nathanss replied:
wait! It works. Give me a couple of minutes

mdperez86 commented on packages/js/components/src/experimental-select-tree-control/stories/index.tsx:
```diff
@@ -37,7 +37,9 @@ const getFilteredItems = ( items: Item[], searchValue ) => {
 		if ( element ) {
-			const parent = listItems.find( ( x ) => x.value === element.parent );
+			const parent = listItems.find(
+				( x ) => x.value === element.parent
```
We should be more explicit here right? `x` -> `item`

mdperez86 reviewed (commented):
Left some comments @nathanss!!! But I think just only one is very important. Good job.

nathanss committed:
Make id mandatory in SelectTree

nathanss committed:
Refactor

nathanss committed:
Improve selector to focus on create button

mdperez86 reviewed (approved):
LGTM! Nice job here @nathanss.
